advanced_source/python_extension_autoload.rst

```diff
@@ -0,0 +1,135 @@
+===========================================
+
+What is it?
+-----------
+
+The extension autoloading mechanism enables PyTorch to automatically
+load out-of-tree backend extensions without explicit import statements. This
+mechanism is very useful for users. On the one hand, it improves the user
+experience and enables users to adhere to the familiar PyTorch device
+programming model without needing to explicitly load or import device-specific
+extensions. On the other hand, it facilitates effortless
+adoption of existing PyTorch applications with zero-code changes on
+out-of-tree devices. For more information,
+see `[RFC] Autoload Device Extension <https://github.com/pytorch/pytorch/issues/122468>`_.
+
+.. note::
+
+    This feature is enabled by default and can be disabled using
+    ``export TORCH_DEVICE_BACKEND_AUTOLOAD=0``.
+    If you get an error like this: "Failed to load the backend extension",
+    this error has nothing to do with PyTorch, you should disable this feature
+    and ask the out-of-tree extension maintainer for help.
+
+How to apply this mechanism to out-of-tree extensions?
+------------------------------------------------------
+
+For example, if you have a backend named ``foo`` and a package named
+``torch_foo``. Make sure your package is based on PyTorch 2.5+ and includes
+the following in its ``__init__.py``:
+
+.. code-block:: python
+
+    def _autoload():
+        print("No need to import torch_foo anymore! Check things are working with `torch.foo.is_available()`.")
+
+Then the only thing you need to do is add an entry point to your Python
+package:
+
+.. code-block:: python
+
+    setup(
+        name="torch_foo",
+        version="1.0",
+        entry_points={
+            "torch.backends": [
+                "torch_foo = torch_foo:_autoload",
+            ],
+        }
+    )
+
+Now the ``torch_foo`` module can be imported when running import torch:
+
+.. code-block:: python
+
+    >>> import torch
+    No need to import torch_foo anymore! Check things are working with `torch.foo.is_available()`.
+    >>> torch.foo.is_available()
+    True
+
+Examples
+^^^^^^^^
+
+Here we take Intel Gaudi HPU and Huawei Ascend NPU as examples to determine how to
+integrate your out-of-tree extension with PyTorch based on the autoloading
+mechanism.
+
+`habana_frameworks.torch`_ is a Python package that enables users to run
+PyTorch programs on Intel Gaudi via the PyTorch ``HPU`` device key.
+``import habana_frameworks.torch`` is no longer necessary after this mechanism
+is applied.
+
+.. _habana_frameworks.torch: https://docs.habana.ai/en/latest/PyTorch/Getting_Started_with_PyTorch_and_Gaudi/Getting_Started_with_PyTorch.html
+
+.. code-block:: diff
+
+    import torch
+    import torchvision.models as models
+    - import habana_frameworks.torch # <-- extra import
+    model = models.resnet50().eval().to("hpu")
+    input = torch.rand(128, 3, 224, 224).to("hpu")
+    output = model(input)
+
```

Comment:
@jgong5 should the implementation example of this mechanism be added here?

Reply:
Yes, please leave it here for now. @bsochack is the habana bridge ready for autoloading?

Reply:
@bsochack we need an implementation example here, can you provide one? Thanks!

Reply:
Any suggestion how it should be provided?

The only challenge that we had is the circular dependencies. We solved it in the way below. Other than that, it worked out of box.

habana_frameworks/__init__.py:
```
import os

is_loaded = False  # A member variable of habana_frameworks module to track if our module has been imported

def __autoload():
    # This is an entrypoint for pytorch autoload mechanism
    # If the following confition is true, that means our backend has already been loaded, either explicitly
    # or by the autoload mechanism and importing it again should be skipped to avoid circular imports
    global is_loaded
    if is_loaded:
        return
    import habana_frameworks.torch
```

habana_frameworks/torch/__init__.py:

```
import os

# This is to prevent torch autoload mechanism from causing circular imports
import habana_frameworks

habana_frameworks.is_loaded = True
```

and one addition in setup.py:
```
entry_points={
        "torch.backends": [
            "device_backend = habana_frameworks:__autoload",
        ],
},
```

Reply:
@bsochack thanks so much for your quick reply!

The circular dependency issue is worth noting, I will state this in this doc.

I noticed habana_frameworks is not an open-source project, so are there any public links about the autoloading mechanism? It's OK if not.

Reply:
Right, habana_frameworks is not open source and there is no official release with the above change.
Btw, it requires PT 2.5 (not yet released) so it is hard to provide a real example.

Reply:
@bsochack OK it's alright

Reply:
@bsochack updated, please take a look at the HPU section

Reply:
lgtm